New Post: how to use transformersjs in a chrome extension (#3361)
huggingface/blog: transformersjs-chrome-extension.md

### 1.1 Runtime contexts and entry points

In MV3, your architecture starts in [`public/manifest.json`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/public/manifest.json). This project defines three entry points:

- `background.service_worker = background.js`, built from [`src/background/background.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/background/background.ts).
- `side_panel.default_path = sidebar.html`, built from [`src/sidebar/index.html`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/sidebar/index.html).
- `content_scripts[].js = content.js` with `matches: http(s)://*/*` and `run_at: document_idle`, built from [`src/content/content.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/content/content.ts).

The background service worker also handles `chrome.action.onClicked` to open the side panel for the active tab.
Related entry point to know: a popup can be defined with `action.default_popup` and works well for quick actions. This project uses a side panel for persistent chat, but the orchestration pattern is the same.

### 1.2 What runs where

> The key design decision is to keep heavy orchestration in the background and keep UI/page logic thin.

Reviewer: maybe explain *why* we want heavy stuff in the background.

Reviewer: just 1 sentence/phrase should do

Author: I think I explained the architectual- and cache-origin-benefits and also chrome specific reasons (security boundaries) already in the next paragraphs. Do you mean I should also add the worker-thread benefits? Or what specifically?

Author: In my opinion its makes sense to focus on the "what" part here. Its more about explaining the the structure and no about explaining all benefits. I briefly touched it in line 56, but thats it.
The full explanation follows in 2.2 where I dive deeper into where the inference runs and above all "why".

Reviewer: sgtm ✅

Typical request flow:

1. Side panel sends `AGENT_GENERATE_TEXT`.
2. Background appends to `Agent.chatMessages` and runs model/tool steps.
3. Background emits `MESSAGES_UPDATE`.
4. Side panel re-renders from the updated message list.

## 2) Transformers.js integration details

### 2.1 Models and responsibilities

In [`src/shared/constants.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/shared/constants.ts), this extension uses two model roles:

- TextGeneration / LLM: [`onnx-community/gemma-4-E2B-it-ONNX`](https://huggingface.co/onnx-community/gemma-4-E2B-it-ONNX) (`text-generation`, `q4f16`)
- VectorEmbeddings: [`onnx-community/all-MiniLM-L6-v2-ONNX`](https://huggingface.co/onnx-community/all-MiniLM-L6-v2-ONNX) (`feature-extraction`, `fp32`)

The split is intentional: Gemma 4 handles reasoning/tool decisions, while MiniLM generates vector embeddings for the semantic similarity search in `ask_website` and `find_history`.

### 2.2 Where inference runs

All inference runs in background ([`src/background/background.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/background/background.ts)):

> - text generation via `pipeline("text-generation", ...)` with consistent `KVCache` enabled by our new `DynamicCache` class

Reviewer: ```suggestion
- text generation via `pipeline("text-generation", ...)` with consistent KV Caching enabled by our new `DynamicCache` class
```

since `KVCache` isn't a keyword used by the lib

- embeddings via `pipeline("feature-extraction", ...)` plus vector normalization

> This gives a single model host for all tabs/sessions, avoids duplicate memory usage, and keeps the side panel UI responsive. Because models are loaded from the background service worker, artifacts are cached under the extension origin (`chrome-extension://<extension-id>`) rather than per-website origins, which gives one shared cache for the whole extension install.

Reviewer: this was basically the explanation i was looking for above. maybe it's fine here, but I think mentioning it earlier could be useful. (or at least, mention a shortened version earlier)

Author: I see. Lets close this and discuss it in the comment above

  - embedding pipeline in [`src/background/utils/FeatureExtractor.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/background/utils/FeatureExtractor.ts)

Permissions and privacy are part of the architecture, not a checkbox at the end. In this project, [`public/manifest.json`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/public/manifest.json) asks for `sidePanel`, `storage`, `scripting`, and `tabs`, plus `host_permissions` for `http(s)://*/*`:

- `sidePanel`: required to open and control the side panel UX.
- `storage`: required to persist tool/settings state across sessions.
- `tabs` + `scripting`: required for tab-aware tools and page-level actions.
- `host_permissions` on `http(s)://*/*`: required because content extraction/highlighting is designed to work on arbitrary websites.

Why keep this narrow: permissions define user trust and Chrome Web Store review risk. Request only what your features actually need, and state clearly that inference runs locally in the extension runtime so users understand where their data is processed.

## 3) Agent and tool execution loop

### 3.1 Tool-calling basics (why this layer exists)

> Before the execution loop, it helps to understand how model tool calling (which is the basis for any agentic workflow) works. You pass messages plus a tool schema (`name`, `description`, and `parameters`), and Transformers.js formats the actual prompt from those inputs using the model's chat template. Because chat templates are model-specific, the exact tool-call format depends on the model you use. With Gemma-4-style templates, the model emits a special tool-call token block when it decides to call one.

Reviewer: ```suggestion
Before the execution loop, it helps to understand how model tool calling works (the basis for any agentic workflow). You pass messages plus a tool schema (`name`, `description`, and `parameters`), and Transformers.js formats the actual prompt from those inputs using the model's chat template. Because chat templates are model-specific, the exact tool-call format depends on the model you use. With Gemma-4-style templates, the model emits a special tool-call token block when it decides to call one.
```

- `name`, `description`, `inputSchema`, `execute`

Example tools include `get_open_tabs`, `go_to_tab`, `open_url`, `close_tab`, `find_history`, `ask_website`, and `highlight_website_element`.

### 3.3 Loop design (`Agent.runAgent`)

The core design choice here is to separate internal model messages from UI-facing chat messages:

- Internal model transcript (`messages`): system/user/tool/assistant turns used for messages in `generator(...)`.
- UI transcript (`chatMessages`): what the user sees, including streamed assistant text plus tool execution metadata (`tools`) and performance metrics.

Execution flow:

1. Add user input to `chatMessages`, create a placeholder assistant message, and stream tokens.
2. Parse streamed/final model output with [`extractToolCalls.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/background/agent/extractToolCalls.ts) into `{ message, toolCalls }`.
3. Keep the user-visible assistant message as plain text, while tool calls execute in background.
4. Append tool results to the assistant tool metadata and feed results back as the next prompt turn.
5. Repeat until no tool calls remain, then finalize assistant content + metrics.

> This keeps user communication clean while preserving a deterministic tool loop in background.

Reviewer: ```suggestion
This keeps user communication clean while preserving a deterministic tool loop in the background.
```

  - [`src/sidebar/index.html`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/sidebar/index.html)
  - [`src/background/background.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/background/background.ts)
  - [`src/content/content.ts`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/src/content/content.ts)
- Ensure manifest-aligned output names/paths (`sidebar.html`, `background.js`, `content.js`).
- Keep the content script as a self-contained output to avoid runtime chunk-loading issues.

The goal is simple: one artifact per Chrome entry point, in the exact place [`public/manifest.json`](https://github.com/nico-martin/gemma4-browser-extension/blob/main/public/manifest.json) expects.

## 6) Debugging checklist

When something breaks, debug by runtime boundary, not by feature:

- Background: verify incoming task messages and model/tool execution logs.
- Side panel: verify `MESSAGES_UPDATE` rendering and loading/error states.
- Model setup: verify first-run `DOWNLOAD_PROGRESS` and warm start from cache (see section 2.3).
- Lifecycle: simulate service worker restart and confirm recovery paths.

> - Content script: validate `EXTRACT_PAGE_DATA` and highlight/clear roundtrip on real pages.

Reviewer: do you think we need to include? or is this more as a warning sign to users for things they should consider when something of theirs (i.e., new changes) breaks?
